torch/distributed/_composable/fsdp/_fsdp_param.py

```diff
@@ -58,11 +58,17 @@ class ShardedState(Enum):
     """
     - ``SHARDED``: The sharded parameter is registered to the module. It is the
       only contributor to parameter memory.
+    - ``SHARDED_POST_FORWARD``: The unsharded parameter is resharded to a
+      smaller world size. Since this data should not be used for computation,
+      we do not register it to the module. Users should reshard the module
```

Comment:
I mentioned briefly with @wz337 and @wanchaol. Perhaps the morally right thing to do here is to still register these tensors but as `DTensor`s over the smaller mesh.

In that case, the implementation gap is mainly for the TP case because today we cannot slice a submesh from an existing mesh and have it preserve its parent mesh (to my knowledge).